presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java

```diff
@@ -1935,6 +1978,37 @@ private PhysicalOperation createLookupJoin(JoinNode node,
             return new PhysicalOperation(operator, outputMappings.build(), context, probeSource);
         }
 
+        private Optional<DynamicFilter> createDynamicFilter(JoinNode node, LocalExecutionPlanContext context)
+        {
+            if (!isEnableDynamicFiltering(context.getSession())) {
+                return Optional.empty();
+            }
+            return DynamicFilter
+                    .create(node, context.getTypes())
+                    .map(filter -> {
```

Comment:
You can use `MoreFutures#addSuccessCallback`.
This can be then simplified to:
```
filter -> addSuccessCallback(filter.getResultFuture(), tuple -> context.intersectDynamicFilter(filter.getTableScanNode(), tuple))
```
assuming future does not return optional and errors are reported and handled in `DynamicFilter`.

Reply:
Actually. Scratch that.
I think you could simply check if `filter.getResultFuture()` is done when you create new `PageSource` for particular table scan.

Reply:
> You can use `MoreFutures#addSuccessCallback`.

Thanks, done at 7b91508809.

Reply:
> I think you could simply check if filter.getResultFuture() is done when you create new PageSource for particular table scan.

This can work, but in order to support multiple join nodes in the same plan fragment (e.g. [TPCDS q13](https://github.com/prestosql/presto/blob/master/presto-product-tests/src/main/resources/sql-tests/testcases/tpcds/q13.sql)), we'll need to intersect several future results for the same TableScan node. 
I have added 972c66d15f for supporting this use-case, WDYT?

Reply:
> Can we assume that there is a single TableScan in each stage?

No. In colocated join there might be two table scans for different tables in same scan.

Reply:
Good point, thanks for the clarification!